### PR TITLE
Update `webp_uploads_pre_generate_additional_image_source` filter to allow returning file size

### LINKED ISSUE #333: Enhance `webp_uploads_pre_generate_additional_image_source` filter to support returning `filesize` value
Follow-up to #307: As identified by @mitogh in https://github.com/WordPress/performance/pull/318#pullrequestreview-971118682 and then further discussed in https://github.com/WordPress/performance/issues/160#issuecomment-1125271238, we should allow developers to optionally return the file size of a custom generated additional image source when using the `webp_uploads_pre_generate_additional_image_source`. This is crucial for implementations where e.g. the image is hosted in a CDN, which would mean that calling `filesize()` on a path would not be feasible.

It is probably best to make the `filesize` field optional, for two reasons:
* To not break any existing filter usage, because the current version of the filter will already be released as part of plugin version 1.1.0.
* To keep things simpler for plugins that _do_ generate files within the WordPress environment, in which case WordPress can just call `filesize()` on the `path` returned.

Since the `path` returned is only relevant for calling `filesize()`, I suggest we implement the change as follows:
* Continue to require the `file` key to be included in the array.
* In addition to that, require _either_ `path` _or_ `filesize` to be included in the array.
* In case of `path`, keep calling `filesize()` on it like today.
* In case of `filesize`, use that value as is to return it within the result array.

### PR DESCRIPTION
**Pull request by eugene-manuilov**
## Summary

Fixes #333

## Checklist

- [x] PR has either `[Focus]` or `Infrastructure` label.
- [x] PR has a `[Type]` label.
- [x] PR has a milestone or the `no milestone` label.

### CODE REVIEW

File: modules/images/webp-uploads/helper.php [hunk 1]
```diff
@@ -84,19 +83,23 @@ function webp_uploads_generate_additional_image_source( $attachment_id, $image_s
...
+			! empty( $image['path'] ) ||
+			! empty( $image['filesize'] )
+		)
 	) {
 		return array(
 			'file'     => $image['file'],
-			'filesize' => filesize( $image['path'] ),
+			'filesize' => ! empty( $image['filesize'] )
+				? $image['filesize']
+				: filesize( $image['path'] ),
```
Comment (mitogh) [inline modules/images/webp-uploads/helper.php]: Probably personal preference but I rather swap the condition here to remove the `!` operator.

```suggestion
			'filesize' => empty( $image['filesize'] )
				? filesize( $image['path'] )
				: $image['filesize'],
```

Another thing (unlikely) is the fact that `empty` would return `false` even when the key is present for `filesize` but the value is `0` or not truthy. I think would be safer to use either `isset` or `array_key_exists` instead.

```suggestion
			'filesize' => isset( $image['filesize'] )
				? $image['filesize']
				: filesize( $image['path'] ),
```

Or

```suggestion
			'filesize' => array_key_exists( $image,'filesize' )
				? $image['filesize']
				: filesize( $image['path'] ),
```

File: modules/images/webp-uploads/helper.php [hunk 2]
```diff
@@ -84,19 +83,23 @@ function webp_uploads_generate_additional_image_source( $attachment_id, $image_s
...
 
 	if (
-		is_array( $image )
-		&& ! empty( $image['file'] )
-		&& ! empty( $image['path'] )
+		is_array( $image ) &&
+		! empty( $image['file'] ) &&
+		(
+			! empty( $image['path'] ) ||
+			! empty( $image['filesize'] )
```
Comment (mitogh) [inline modules/images/webp-uploads/helper.php]: Same comments apply here for the scenario where the key is present but contains a falsy value, would be safer to go with `isset` or `array_key_exists`

File: modules/images/webp-uploads/helper.php (hunk 1 above)
Comment (eugene-manuilov) [reply modules/images/webp-uploads/helper.php]: Yep. Makes sense. Updated.

File: modules/images/webp-uploads/helper.php (hunk 2 above)
Comment (eugene-manuilov) [reply modules/images/webp-uploads/helper.php]: Updated.

File: modules/images/webp-uploads/helper.php [hunk 3]
```diff
@@ -84,19 +83,23 @@ function webp_uploads_generate_additional_image_source( $attachment_id, $image_s
...
+			! empty( $image['path'] ) ||
+			array_key_exists( 'filesize', $image )
+		)
 	) {
 		return array(
 			'file'     => $image['file'],
-			'filesize' => filesize( $image['path'] ),
+			'filesize' => array_key_exists( 'filesize', $image )
+				? $image['filesize']
+				: filesize( $image['path'] ),
```
Comment (mehulkaklotar) [inline modules/images/webp-uploads/helper.php:102]: We are also returning same array in later in this function. https://github.com/WordPress/performance/pull/334/files#diff-1f653bd37f60a6476485a6794f7003a91b3713868c7414588d5835332fe68642R156

so can we put checks there as well?

**Review by mehulkaklotar — CHANGES_REQUESTED** [review]

File: modules/images/webp-uploads/helper.php (hunk 3 above)
Comment (eugene-manuilov) [reply modules/images/webp-uploads/helper.php:102]: That is different. The `$image` variable is returned from the image editor which doesn't return file size information, thus we don't need to check for the `filesize` property there.

File: tests/modules/images/webp-uploads/helper-tests.php [hunk 4]
```diff
@@ -270,7 +270,33 @@ function () {
...
 		$this->assertStringEndsWith( 'image.webp', $result['file'] );
-		$this->assertFileExists( '/tmp/image.webp' );
+	}
+
+	/**
+	 * Tests the webp_uploads_pre_generate_additional_image_source filter returning filesize property.
+	 *
+	 * @test
+	 */
+	public function it_should_return_an_error_when_filter_webp_uploads_pre_generate_additional_image_source_returns_filesize() {
```
Comment (felixarntz) [inline tests/modules/images/webp-uploads/helper-tests.php]: This name is not quite right as far as I can tell? Where in this test is there any error to be returned?

**Review by felixarntz — COMMENTED** [review]
@eugene-manuilov Looks pretty much good to go, just one tiny comment.

File: tests/modules/images/webp-uploads/helper-tests.php (hunk 4 above)
Comment (eugene-manuilov) [reply tests/modules/images/webp-uploads/helper-tests.php]: Good catch. Forgot to change that part of the name after copying the test with an error. Now it should be good, do you mind checking it again?

File: tests/modules/images/webp-uploads/helper-tests.php [hunk 5]
```diff
@@ -270,7 +270,33 @@ function () {
...
+			'webp_uploads_pre_generate_additional_image_source',
+			function () {
+				return array(
+					'file'     => 'image.webp',
+					'filesize' => 777,
+				);
+			}
+		);
+
+		$result = webp_uploads_generate_additional_image_source( 0, 'medium', array(), 'image/webp', '/tmp/image.jpg' );
```
Comment (felixarntz) [inline tests/modules/images/webp-uploads/helper-tests.php]: Not a blocker, but I'd be in favor of using a real `$attachment_id` and `$size_data` here, similar to the test method right above. Yes, it is strictly not necessary, but I think it's better to use data that is appropriate for what the function expects. Unless we are testing explicitly how it behaves with "weird" data, which is not the case here though.

**Review by felixarntz — APPROVED** [review]
@eugene-manuilov Looks good!

One remaining non-blocking comment.

**Review by jjgrainger — APPROVED** [review]

File: tests/modules/images/webp-uploads/helper-tests.php (hunk 5 above)
Comment (eugene-manuilov) [reply tests/modules/images/webp-uploads/helper-tests.php]: Ok, updated.

**Review by felixarntz — APPROVED** [review]

**Review by mehulkaklotar — APPROVED** [review]
LGTM 🚀

**Pull request merged**